### PR TITLE
🔥 feat: Add SPNEGO Authentication Middleware

### LINKED ISSUE #1367: 📝 [Proposal]: Add Enterprise-Grade SPNEGO Authentication Middleware
### Feature Proposal Description

**Add Enterprise-Grade SPNEGO Authentication Middleware**  
This middleware provides Kerberos-based authentication for Fiber v3:
- Handles full SPNEGO protocol (`WWW-Authenticate` challenges + `Authorization` validation)
- Dynamic keytab loading via callback functions (files, secrets managers, databases)
- Sets authenticated principal in `c.Locals("user")`
- System credential cache fallback
- Automatic SPN/realm resolution
- Customizable unauthorized responses

**Workflow**:
1. Client accesses protected route
2. Middleware returns `401 + WWW-Authenticate: Negotiate`
3. Client provides Kerberos ticket (browser auto-completes)
4. Middleware validates ticket and decrypt request user info
5. Success: Continue route with `c.Locals("user")` set

### Alignment with Express API

This implementation follows Express patterns
```go
authMiddleware, err := NewSpnegoKrb5AuthenticateMiddleware(&Config{
	KeytabLookup: keytabLookup,
})
app.Use("/protected", authMiddleware)
app.Get("/protected/resource", func(c fiber.Ctx) error {
	identity, ok := GetAuthenticatedIdentityFromContext(c)
	if !ok {
		return c.Status(fiber.StatusUnauthorized).SendString("Unauthorized")
	}
	return c.SendString(fmt.Sprintf("Hello, %s!", identity.UserName()))
})
```

### HTTP RFC Standards Compliance

Fully compliant with:
- **RFC 4559** (SPNEGO-based Kerberos HTTP Authentication)
  - Section 4.1: `Negotiate` scheme implementation
  - Section 4.2: GSS-API message encapsulation
- **RFC 4120** (Kerberos V5)
  - AP-REQ/AP-REP validation
  - Service ticket verification
- **RFC 7235** (HTTP/1.1 Authentication)
  - `WWW-Authenticate` challenge generation
  - `Authorization` header processing

**Compliance ensured through**:
- Standardized [gokrb5](https://github.com/jcmturner/gokrb5) implementation
- Test coverage for all authentication flows
- Support for MIT Kerberos and Active Directory

### API Stability

```go

// KeytabLookupFunc is a function type that returns a keytab or an error
// It's used to look up the keytab dynamically when needed
// This design allows for extensibility, enabling keytab retrieval from various sources
// such as databases, remote services, or other custom implementations beyond static files
type KeytabLookupFunc func() (*keytab.Keytab, error)

// Config holds the configuration for the SPNEGO middleware
// It includes the keytab lookup function and a logger
type Config struct {
	// KeytabLookup is a function that retrieves the keytab
	KeytabLookup KeytabLookupFunc
	// Log is the logger used for middleware logging
	Log *log.Logger
}
```

### Feature Examples

```go
This just a example server
1. user should use `kinit` login to kerberos at first and use `klist` check session.
2. use `curl --negotiate ` to trigger spnego authenticate

app := fiber.New()
keytabLookup, err := NewKeytabFileLookupFunc("/keytabFile/one.keytab", "/keytabFile/two.keyta")
if err != nil {
	panic(fmt.Errorf("create keytab lookup function failed: %w", err))
}
authMiddleware, err := NewSpnegoKrb5AuthenticateMiddleware(&Config{
	KeytabLookup: keytabLookup,
})
if err != nil {
	panic(fmt.Errorf("create spnego middleware failed: %w", err))
}
// Apply the middleware to protected routes
app.Use("/protected", authMiddleware)
// Access authenticated identity
app.Get("/protected/resource", func(c fiber.Ctx) error {
    identity, ok := GetAuthenticatedIdentityFromContext(c)
	if !ok {
		return c.Status(fiber.StatusUnauthorized).SendString("Unauthorized")
	}
	return c.SendString(fmt.Sprintf("Hello, %s!", identity.UserName()))
})
log.Info("Server is running on :3000")
go func() {
	<-time.After(time.Second * 1)
	fmt.Println("use curl -kv --negotiate http://sso.example.local:3000/protected/resource")
	fmt.Println("if response is 401, execute `klist` to check use kerberos session")
	<-time.After(time.Second * 2)
	fmt.Println("close server")
	if err = app.Shutdown(); err != nil {
		panic(fmt.Errorf("shutdown server failed: %w", err))
	}
}()
if err := app.Listen("sso.example.local:3000"); err != nil {
	panic(fmt.Errorf("start server failed: %w", err))
}
```

### Checklist:

- [x] I agree to follow Fiber's [Code of Conduct](https://github.com/gofiber/fiber/blob/main/.github/CODE_OF_CONDUCT.md).
- [x] I have searched for existing issues that describe my proposal before opening this one.
- [x] I understand that a proposal that does not meet these guidelines may be closed without explanation.

### PR DESCRIPTION
**Pull request by jarod2011**
## Description
This PR adds enterprise-grade SPNEGO authentication middleware to Fiber, providing Kerberos-based single sign-on (SSO) capabilities. The implementation follows Fiber's middleware patterns with dynamic keytab loading and full RFC compliance.

## Related Issue
Resolves #1367  
Issue: [SPNEGO Middleware Proposal](https://github.com/gofiber/contrib/issues/1367)

## Technical Implementation
- 🛡️ **SPNEGO Protocol**: Full support for `WWW-Authenticate` challenges and `Authorization` validation
- 🔑 **Dynamic Credentials**: Keytab loading via callback functions (files, secrets managers, databases)
- ✅ **RFC Compliance**: Fully implements RFC 4559, RFC 4120, and RFC 7235
- 📊 **Test Coverage**: >90% coverage including edge cases and error handling
- 📚 **Documentation**: Complete usage examples in README
- Support fiber v2 and v3

## Verification
```bash
# Run tests
cd spnego
go test -cover ./...
# Test output
        github.com/gofiber/contrib/spnego               coverage: 0.0% of statements
        github.com/gofiber/contrib/spnego/config                coverage: 0.0% of statements
ok      github.com/gofiber/contrib/spnego/v2    0.389s  coverage: 87.5% of statements
ok      github.com/gofiber/contrib/spnego/v3    0.388s  coverage: 87.5% of statements


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * SPNEGO Kerberos authentication middleware for Fiber v3 with configurable keytab lookup and optional logging
  * Context helpers to store/retrieve authenticated identity

* **Documentation**
  * New README, package docs, and end-to-end example demonstrating setup and usage

* **Utilities**
  * Keytab inspection helpers and a mock keytab builder for tests/examples

* **Tests**
  * Extensive unit and integration tests covering keytab lookup, middleware flows, and utilities
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

### CODE REVIEW

File: spnego/README.md [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+This middleware is available in two versions to support different Fiber releases:
+
+- **v2**: Compatible with Fiber v2
+- **v3**: Compatible with Fiber v3
+
+## Installation
+
+```bash
+# For Fiber v3
+go get github.com/gofiber/contrib/spnego/v3
```
Comment (ReneWerner87) [inline spnego/README.md]: v2: github.com/gofiber/contrib/v2/spnego
v3: github.com/gofiber/contrib/v3/spnego

You can arrange it so that the packet itself can also have major versions in the future.
future : github.com/gofiber/contrib/v3/spnego/v2

Of course, the go.mod would have to be in the subfolders for that.

File: spnego/README.md [hunk 2]
```diff
@@ -0,0 +1,197 @@
+# SPNEGO Kerberos Authentication Middleware for Fiber
```
Comment (ReneWerner87) [inline spnego/README.md]: pls follow our default structure and add a header like this

Image

Comment (jarod2011) [reply spnego/README.md]: ok,i add a header to my README.md

File: spnego/README.md (hunk 1 above)
Comment (ReneWerner87) [reply spnego/README.md]: @jarod2011 We are currently working on this change and plan to use the prefix.
Can you please adjust it accordingly so that it fits into the scheme?

Comment (jarod2011) [reply spnego/README.md]: Hi  @ReneWerner87 ,

Thanks for your feedback!

To make sure I understand correctly and to align with the new scheme: should I split this PR into two?

    One for the Fiber v3 compatibility, which would be merged into the v3-beta branch.

    And another one for the Fiber v2 compatibility, which would be merged into the v2 branch.

Please let me know if this is the right approach. I'm happy to adjust accordingly.

Thanks!

**Comment by gaby** [comment]
@jarod2011 Few small things:
- You are missing a test workflow in .github/workflows/ copy one of the existing ones and change whatever needs to be change
- Remove the adapter util, we do not need to provide support for v2 at this point
- Remove `v3/spnego/README.zh-CN.md`. We only support English README files.

**Comment by gaby** [comment]
@jarod2011 Few small things:
- Missing test-spnego.yml workflow
- We do not need the adaptor util
- The example has several issues as mentioned above
- For creating the middleware the function should be named `New()`.

File: v3/spnego/config.go [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+// such as databases, remote services, or other custom implementations beyond static files
+type KeytabLookupFunc func() (*keytab.Keytab, error)
+
+// Config holds the configuration for the SPNEGO middleware
+// It includes the keytab lookup function and a logger
+type Config struct {
+	// KeytabLookup is a function that retrieves the keytab
+	KeytabLookup KeytabLookupFunc
+	// Log is the logger used for middleware logging
+	Log *log.Logger
```
Comment (gaby) [inline v3/spnego/config.go]: We should use fiberlog here. 

https://docs.gofiber.io/next/api/log#global-log
